sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -361,6 +361,18 @@ object KeyGroupedPartitioning {
       partitionValues: Seq[InternalRow]): KeyGroupedPartitioning = {
     KeyGroupedPartitioning(expressions, partitionValues.size, Some(partitionValues))
   }
+
+  def supportsExpressions(expressions: Seq[Expression]): Boolean = {
+    def isSupportedTransform(transform: TransformExpression): Boolean = {
+      transform.children.size == 1 && transform.children.head.isInstanceOf[AttributeReference]
+    }
+
+    expressions.forall {
+      case _: NamedExpression => true
```

Comment:
`NamedExpression` is not `NamedReference`. e.g. `a + b AS c` is also a `NamedExpression`. We can add a small util function to check nested columns

```
def isReference(e: Expression): Boolean = e match {
  case a: Attribute => true
  case g: GetStructField => isReference(g.child)
  case _ => false
}
```

Reply:
I was mainly judging from the perspective of `V2ExpressionUtils` but I agree this is safer. Changed.